src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,399 @@
+        // A switch would be nicer, but we may need to backport this before
+        // JDK 17
+        // Also, JEP 305 came out in JDK 14, so we can't declare a variable
+        // in instanceof either
+        if (derivationParameterSpec instanceof HKDFParameterSpec.Extract) {
+            HKDFParameterSpec.Extract anExtract =
+                (HKDFParameterSpec.Extract) derivationParameterSpec;
+            ikms = anExtract.ikms();
+            salts = anExtract.salts();
+            // we should be able to combine these Lists of keys into single
+            // SecretKey Objects,
+            // unless we were passed something bogus or an unexportable P11 key
+            try {
+                inputKeyMaterial = consolidateKeyMaterial(ikms);
+                salt = consolidateKeyMaterial(salts);
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "Issue encountered when combining ikm or salt values into"
+                    + " single keys").initCause(ike);
+            }
+            // perform extract
+            try {
+                return hkdfExtract(inputKeyMaterial,
+                                   (salt == null) ? null : salt.getEncoded());
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "an HKDF Extract could not be initialized with the given "
+                    + "key or salt material").initCause(ike);
+            } catch (NoSuchAlgorithmException nsae) {
+                // This is bubbling up from the getInstance of the Mac/Hmac.
+                // Since we're defining these values internally, it is unlikely.
+                throw new ProviderException(
+                    "could not instantiate a Mac with the provided algorithm",
+                    nsae);
+            }
+        } else if (derivationParameterSpec instanceof HKDFParameterSpec.Expand) {
+            HKDFParameterSpec.Expand anExpand =
+                (HKDFParameterSpec.Expand) derivationParameterSpec;
+            // set this value in the "if"
+            if ((pseudoRandomKey = anExpand.prk()) == null) {
+                throw new InvalidAlgorithmParameterException(
+                    "PRK is required for HKDFParameterSpec.Expand");
+            }
+            // set this value in the "if"
+            if ((info = anExpand.info()) == null) {
+                info = new byte[0];
+            }
+            length = anExpand.length();
+            if (length > (hmacLen * 255)) {
+                throw new InvalidAlgorithmParameterException(
+                    "Requested length exceeds maximum allowed key stream "
+                    + "length");
+            }
+            // perform expand
+            try {
+                return Arrays.copyOf(hkdfExpand(pseudoRandomKey, info, length),
```

Comment:
If the requested `length` happens to be multiples of `hmacLen`, then we can return the output of `hkdfExpand()` directly. No need for `Arrays.copyOf()`.

Reply:
Since the desired length is passed into `hkdfExpand()` method, why not make `hkdfExpand()` return the output with the requested length?

Reply:
Not entirely addressed. Please see new comments.